お部屋のスコア編集ページを作成する

uchihiro04 opened the pull request:
## 概要 
お部屋のスコア編集ページを作成しました。
あわせて、スコア入力・編集ページの「入力者一覧」について、入力者ごとに「編集」ボタンのリンクを設定しました。

## スクリーンショット
* お部屋のスコア編集ページ
<img width="700" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/ebbd41ee-b295-4665-9813-62aee5e0f1ca">

デベロッパツールで「iPhone SE」の画面（375 × 667）を想定して表示した場合
<img width="200" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/e562123a-de76-4972-b3a6-c9c42071761b">


* 作成後の画面
お部屋の内見一覧ページにリダイレクトし、「スコアを登録しました」と表示されます（赤枠の部分）。
<img width="700" alt="スクリーンショット 2023-06-30 9 55 56" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/b4f6eecc-590e-403e-866c-36b3915bc80d">

デベロッパツールで「iPhone SE」の画面（375 × 667）を想定して表示した場合（赤枠の部分）。
<img width="200" alt="スクリーンショット 2023-06-30 9 58 46" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/1e320716-180f-40d0-91db-e029b0e5bdf0">

## 関連Issue
- #82 

## 備考
以下の実装は別Issueで行います。
* デザイン

Close #82

uchihiro04 committed:
#82 スコア編集ページを表示する

スコア編集ページを表示しました。現時点では「スコアを保存する」ボタンを押してもスコアは更新されません。

uchihiro04 committed:
#82 スコア編集ページへのリンクを設定

スコア入力者一覧について、スコア編集ページへのリンクを設定しました。

uchihiro04 committed:
#82 updateアクションを追加

uchihiro04 committed:
#82 createアクションでの不具合を修正

createアクション実行時にバリデーションに失敗した際、入力者一覧でエラーが出る不具合を修正しました。

uchihiro04 committed:
82 Scoreコントローラのリファクタリング

Scoreコントローラについて、DRYになっていないコードをリファクタリングしました。

Yuma-Nagaoka commented on app/views/house_viewings/rooms/scores/_form.html.slim:
```diff
@@ -14,7 +14,8 @@
       .flex
         li = score.reviewer_name
-        = link_to '編集'
+        - if score.persisted?
```
これscore.persisted?がtrueにならない状況ってありますか？

uchihiro04 replied:
状況としては、スコアの新規入力時にバリデーションに失敗する形で保存を試みると以下のエラーが発生します。
<img width="700" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/3043d3ab-7682-4df6-bdb5-d06e1696ddac">

`@room`に紐づけられた`scores`の中に、idがnilのScoreレコードが含まれているため、エラーが発生していると考えています。
idがnilのScoreレコードは、DBには保存されていないため`persisted?`メソッドで除外していた形です。別の方法として、Scoreコントローラのcreateアクションにて、`@score.save`を`@score.save!`に変えてもエラーは解消できました。

Scoreコントローラのcreateアクションに関しては、#112 で修正を行っているためコードが変更され、`persisted?`メソッドを使う必要がなくなる可能性もありそうです。

Yuma-Nagaoka replied:
@uchihiro04 使わなくて良くなりそうならそうしましょう！

uchihiro04 replied:
夕会での話し合いをもとに、persisted?メソッドの位置を変更しました。該当するコミットは[d29776c](https://github.com/uchihiro04/house-viewing-score-log/pull/121/commits/d29776ce23d596e65dc12fcfa0b59823e031092f)です。

Yuma-Nagaoka commented on app/views/house_viewings/rooms/scores/edit.html.slim:
```diff
@@ -0,0 +1,5 @@
+.flex
+  = link_to '<', house_viewing_rooms_path
+  h1
+    = "#{@room.name}を編集する"
```
"#{@room.name}を編集する"ための「<」 リンクかと思ったら、前ページの「お部屋の内見一覧ページ」のリンクでした。勘違いしちゃったので一応コメントに残しておきます。

uchihiro04 committed:
#82 persisted?メソッド、ulの位置を修正

条件分岐の範囲を修正するためにpersisted?メソッドの位置を変更しました。また、ulの位置を修正しました。

uchihiro04 committed:
#82 editに関するビューを修正

editに関するビューについて、見出しを修正しました。また、renderにlocalsオプションを追加しました。

uchihiro04 committed:
#82 updateアクションを修正

updateアクションについて、Scoreレコードを更新した際にRoomレコードも更新される形に変更しました。

uchihiro04 committed:
#82 エラーメッセージの不具合を修正

エラーメッセージについて、二重でエラーメッセージが表示される不具合を修正しました。

uchihiro04 committed:
#82 editに関するビューの見出しを修正

uchihiro04 committed:
#82 フラッシュメッセージを国際化対応する

フラッシュメッセージについてrubocopの指摘を受けたため、国際化対応しました。

uchihiro04 committed:
#82 before_actionのメソッドを追加

部屋名の取得についてメソッド化し、before_actionに追加しました。

Yuma-Nagaoka commented on app/controllers/house_viewings/rooms/scores_controller.rb:
```diff
@@ -3,25 +3,37 @@
-        room_name = params[:score][:room][:name]
         @score = @room.scores.new(score_params)
-        @room.name = room_name if @room.name != room_name
+        @room.name = @room_name if @room.name != @room_name
```
~@room_nameについてですが、@new_room_nameにした方がいいと思いました。
@room.nameと@room_nameの違いがわかりにくいため。~

下記コメントで対応すればここは大丈夫ですね
https://github.com/uchihiro04/house-viewing-score-log/pull/121#discussion_r1250303013

また、この部分のif分岐あまりいらないかなと思いました。名前が変わっていなければ同じ名前で再代入するだけなので問題ない気がします。それより「room」がたくさん並ぶことで見辛くなるのを避けたいです。

uchihiro04 replied:
確かに条件分岐がない状態でも問題はなさそうですね……。また、`@room.name`と`@room_name`が並ぶと可読性は落ちるように感じました。

いただいたレビューをもとに修正を行いました。該当するコミットは[c35dbbf](https://github.com/uchihiro04/house-viewing-score-log/pull/121/commits/c35dbbf1830ca54e595301cf2dcaa61ba769e827)です。

Yuma-Nagaoka commented on app/controllers/house_viewings/rooms/scores_controller.rb:
```diff
@@ -3,25 +3,37 @@
-      before_action :set_room, only: %i[new create]
+      before_action :set_house_viewing, :set_room, only: %i[new create edit update]
+      before_action :set_score, only: %i[edit update]
+      before_action :fetch_room_name, only: %i[create update]
```
set_new_room_nameというメソッドにしてしまい、
```
def set_new_room_name
@room.name = params[:score][:room][:name]
end
```
でいいと思います。
fetchだとどこから(DB?パラメータ？)取ってきてるのかわからない気がしました。

uchihiro04 replied:
いただいたレビューをもとに、メソッド名を`set_new_room_name`に変更しました。また、`@room.name`に paramsを代入する形に変更しました。該当するコミットは[c35dbbf](https://github.com/uchihiro04/house-viewing-score-log/pull/121/commits/c35dbbf1830ca54e595301cf2dcaa61ba769e827)です。

uchihiro04 committed:
#82 メソッド名の修正・条件分岐の削除

メソッド名をfetch_room_nameからset_new_room_nameに変更しました。また、条件分岐を削除しました。

Yuma-Nagaoka commented on app/views/house_viewings/rooms/scores/_form.html.slim:
```diff
@@ -1,8 +1,7 @@
-       li = message
+- if score.errors.present?
+  ul
+    - score.errors.full_messages.each do |message|
```
room, scoreのどちらで起きたバリデーションエラーもscore.errorsに入っているという認識であっていますか？

uchihiro04 replied:
コメントいただいた通り、roomで起きたバリデーションエラーもscore.errorsに入っている形になります。
room, scoreのバリデーションエラーを発生させると以下の通りとなり、両方のモデルに関するエラーメッセージが表示されます。

<img width="200" alt="image" src="https://github.com/uchihiro04/house-viewing-score-log/assets/77523896/fe263a2c-9b86-402e-ae61-4b1efeb601ab">

Yuma-Nagaoka replied:
なるほどです！保存処理が@room.saveなのでroom.errorsで出した方がいいかなとも思いましたが、一旦このままで大丈夫です！

uchihiro04 committed:
#82 共通テンプレートについて、ローカル変数を使用する形に変更

フォームに関する共通テンプレートについて、インスタンス変数からローカル変数を使用する形に変更しました。

takuminasu commented on app/controllers/house_viewings/rooms/scores_controller.rb:
```diff
@@ -3,25 +3,34 @@
-      before_action :set_room, only: %i[new create]
+      before_action :set_house_viewing, :set_room, only: %i[new create edit update]
+      before_action :set_score, only: %i[edit update]
+      before_action :set_new_room_name, only: %i[create update]
```
before_actionの乱用はあまりしたくないですね。
#createと#updateで`set_room_name`メソッドを呼び出すようにしたいです。
https://blog.recruit.co.jp/rmp/server-side/rails-development-policy/

uchihiro04 replied:
レビューおよび記事を読み、before_actionを使いすぎるとロジックの把握が困難になり、バグの原因にもなることを把握しました。

レビューをもとにbefore_actionを使用しない形に修正しました。該当するコミットは[cd55ed5](https://github.com/uchihiro04/house-viewing-score-log/pull/121/commits/cd55ed5bf513dd6af4fa9e307ae28b744dd26208)、[ee480af](https://github.com/uchihiro04/house-viewing-score-log/pull/121/commits/ee480af6471085f1aaaa9ba6afbb7b5548349508)です。

uchihiro04 committed:
#82 set_new_room_nameメソッドをbefore_actionから削除

ロジックの把握が難しくなるため、set_new_room_nameメソッドをbefore_actionから削除し、各アクションで呼び出す形に変更しました。

uchihiro04 committed:
#82 メソッド名を変更

set_new_room_nameメソッドについて、set_room_nameに名前を変更しました。理由としては、常に新しい部屋名が入るわけではないため、newをつけない方がいいのではと判断したためです。

takuminasu commented on app/controllers/house_viewings/rooms/scores_controller.rb:
```diff
@@ -3,25 +3,35 @@
 
         if @room.save
-          redirect_to house_viewing_rooms_path, notice: 'スコアを登録しました。'
+          redirect_to house_viewing_rooms_path, notice: t('notice.create', model: @score.class.model_name.human)
```
classがなくてもかけます。
```ruby
@score.model_name.human
=> "スコア"
```

uchihiro04 replied:
確かに`@score.model_name.human`でも同じ結果になりますね……。

いただいたレビューをもとに、`model: @score.class.model_name.human`を`model: @score.model_name.human`に修正しました。該当するコミットは[0f1c15f](https://github.com/uchihiro04/house-viewing-score-log/pull/121/commits/0f1c15f2ff38825729836866ca7fe905f0e473df)です。

uchihiro04 committed:
#82 model_nameメソッドのレシーバを修正

model_nameメソッドのレシーバを`@score.class`から`@score`に修正しました。
